pkg/cloud/azure/actuators/machineset/controller_test.go

```diff
@@ -0,0 +1,146 @@
+		vmSize              string
+		existingAnnotations map[string]string
+		expectedAnnotations map[string]string
+		expectErr           bool
+	}{
+		{
+			name:                "with no vmSize set",
+			vmSize:              "",
+			existingAnnotations: make(map[string]string),
+			expectedAnnotations: make(map[string]string),
+			expectErr:           true,
+		},
+		{
+			name:                "with a Standard_D4s_v3",
+			vmSize:              "Standard_D4s_v3",
+			existingAnnotations: make(map[string]string),
+			expectedAnnotations: map[string]string{
+				cpuKey:    "4",
+				memoryKey: "16384",
+				gpuKey:    "0",
+			},
+			expectErr: false,
+		},
+		{
+			name:                "with a Standard_NC24",
+			vmSize:              "Standard_NC24",
+			existingAnnotations: make(map[string]string),
+			expectedAnnotations: map[string]string{
+				cpuKey:    "24",
+				memoryKey: "229376",
+				gpuKey:    "4",
+			},
+			expectErr: false,
+		},
+		{
+			name:   "with existing annotations",
+			vmSize: "Standard_NC24",
+			existingAnnotations: map[string]string{
+				"existing": "annotation",
+				"annother": "existingAnnotation",
+			},
+			expectedAnnotations: map[string]string{
+				"existing": "annotation",
+				"annother": "existingAnnotation",
+				cpuKey:     "24",
+				memoryKey:  "229376",
+				gpuKey:     "4",
+			},
+			expectErr: false,
+		},
+		{
+			name:   "with an invalid vmSize",
+			vmSize: "r4.xLarge",
+			existingAnnotations: map[string]string{
+				"existing": "annotation",
+				"annother": "existingAnnotation",
+			},
+			expectedAnnotations: map[string]string{
+				"existing": "annotation",
+				"annother": "existingAnnotation",
+			},
+			expectErr: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(tt *testing.T) {
+			g := NewWithT(tt)
+
+			machineSet, err := newTestMachineSet(tc.vmSize, tc.existingAnnotations)
+			g.Expect(err).ToNot(HaveOccurred())
+
+			_, err = reconcile(machineSet)
+			g.Expect(err != nil).To(Equal(tc.expectErr))
+			g.Expect(machineSet.Annotations).To(Equal(tc.expectedAnnotations))
+		})
+	}
+}
+
+func newTestMachineSet(vmSize string, existingAnnotations map[string]string) (*machinev1.MachineSet, error) {
+	// Copy anntotations map so we don't modify the input
+	annotations := make(map[string]string)
+	for k, v := range existingAnnotations {
+		annotations[k] = v
+	}
+
+	machineProviderSpec := &machineproviderv1.AzureMachineProviderSpec{
+		VMSize: vmSize,
+	}
+	providerSpec, err := providerSpecFromMachine(machineProviderSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	return &machinev1.MachineSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations: annotations,
+		},
+		Spec: machinev1.MachineSetSpec{
+			Template: machinev1.MachineTemplateSpec{
+				Spec: machinev1.MachineSpec{
+					ProviderSpec: providerSpec,
+				},
+			},
+		},
+	}, nil
+}
+
+func providerSpecFromMachine(in *machineproviderv1.AzureMachineProviderSpec) (machinev1.ProviderSpec, error) {
```

Comment:
Is there any exported method in the actuator machine controller we can reuse for this?

Reply:
Not exported, but it does already exist in the tests https://github.com/openshift/cluster-api-provider-azure/blob/d80999fe3ae8103510c3239f31f438bee3439142/pkg/cloud/azure/actuators/machine_scope_test.go#L42-L50 and https://github.com/openshift/cluster-api-provider-azure/blob/d80999fe3ae8103510c3239f31f438bee3439142/pkg/cloud/azure/actuators/machine/actuator_test.go#L69-L77